File: Exp3/04_statistical_analysis/Exp3_analyse.py
import os
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import statsmodels.formula.api as smf

# ---------------------------
# Paths
# ---------------------------
script_dir = os.path.dirname(os.path.abspath(__file__))
exp_root = os.path.abspath(os.path.join(script_dir, ".."))

# ---------------------------
# Helper function
# ---------------------------
def compute_ci(df):
    grouped = df.groupby(["Story", "Emotion"])
    out = grouped["Val"].agg(['mean', 'std']).reset_index()
    out.rename(columns={"mean": "Val", "std": "sd"}, inplace=True)

    n = df["ID"].nunique()
    out["SE"] = out["sd"] / np.sqrt(n)
    out["CI_lower"] = np.maximum(out["Val"] - 1.96 * out["SE"], 0)
    out["CI_upper"] = np.minimum(out["Val"] + 1.96 * out["SE"], 1)

    return out[["Story", "Emotion", "Val", "CI_lower", "CI_upper"]]

# ===========================
# 🔵 FREE RATING
# ===========================
data = pd.read_csv(os.path.join(exp_root, "data", "human_free_limit.csv"), sep=";")

data = data.melt(
    id_vars=["ID", "Story"],
    value_vars=[c for c in data.columns if c.startswith("Emo.")],
    var_name="Emotion",
    value_name="Val"
)

data["Emotion"] = data["Emotion"].str.replace("Emo.", "")

# Mixed model
model = smf.mixedlm("Val ~ Story * Emotion", data, groups=data["ID"])
print(model.fit().summary())

# Human CI
data_h = compute_ci(data)
data_h["Source"] = "Human"

# Model data
data_m = pd.read_csv(os.path.join(exp_root, "data", "svm_free_0.0013_var.csv"))
data_m = data_m.groupby(["Story", "Emotion"])["Val"].mean().reset_index()
data_m["CI_lower"] = data_m["Val"]
data_m["CI_upper"] = data_m["Val"]
data_m["Source"] = "Model"

# Combine
data_b = pd.concat([data_h, data_m])

# Plot
g = sns.catplot(
    data=data_b,
    x="Emotion",
    y="Val",
    hue="Source",
    col="Story",
    kind="bar",
    col_wrap=1,
    height=3,
    aspect=2
)

for ax, story in zip(g.axes.flatten(), data_b["Story"].unique()):
    subset = data_b[(data_b["Story"] == story) & (data_b["Source"] == "Human")]
    ax.errorbar(
        x=np.arange(len(subset)),
        y=subset["Val"],
        yerr=[subset["Val"] - subset["CI_lower"], subset["CI_upper"] - subset["Val"]],
        fmt='none',
        capsize=3
    )

plt.ylim(0, 0.6)
plt.tight_layout()
plt.savefig(os.path.join(exp_root, "plots", "exp3_free.png"), dpi=300)

# RMSE
data_bl = data.groupby(["Story", "Emotion"])["Val"].mean().reset_index()
data_bl["Val"] = data_bl.groupby("Story")["Val"].transform(lambda x: x / x.sum())

data_bl = data_bl.merge(data_m, on=["Story", "Emotion"], suffixes=("_x", "_y"))
data_bl["d"] = (data_bl["Val_x"] - data_bl["Val_y"])**2

print("FREE RMSE:", np.sqrt(data_bl["d"].mean()))

# ===========================
# 🔴 LIMIT RATING
# ===========================
data2 = pd.read_csv(os.path.join(exp_root, "data", "human_free_limit.csv"), sep=";")

data2 = data2.melt(
    id_vars=["ID", "Story"],
    value_vars=[c for c in data2.columns if c.startswith("mc")],
    var_name="Emotion",
    value_name="Val"
)

data2["Emotion"] = data2["Emotion"].str.replace("mc.", "")

# Mixed model
model2 = smf.mixedlm("Val ~ Story * Emotion", data2, groups=data2["ID"])
print(model2.fit().summary())

# Human CI
data_h2 = compute_ci(data2)
data_h2["Source"] = "Human"

# Model data
data_m2 = pd.read_csv(os.path.join(exp_root, "data", "svm_limit_0.0034_var.csv"))
data_m2 = data_m2.groupby(["Story", "Emotion"])["Val"].mean().reset_index()
data_m2["CI_lower"] = data_m2["Val"]
data_m2["CI_upper"] = data_m2["Val"]
data_m2["Source"] = "Model"

# Combine
data_b2 = pd.concat([data_h2, data_m2])

# Plot
g2 = sns.catplot(
    data=data_b2,
    x="Emotion",
    y="Val",
    hue="Source",
    col="Story",
    kind="bar",
    col_wrap=1,
    height=3,
    aspect=2
)

for ax, story in zip(g2.axes.flatten(), data_b2["Story"].unique()):
    subset = data_b2[(data_b2["Story"] == story) & (data_b2["Source"] == "Human")]
    ax.errorbar(
        x=np.arange(len(subset)),
        y=subset["Val"],
        yerr=[subset["Val"] - subset["CI_lower"], subset["CI_upper"] - subset["Val"]],
        fmt='none',
        capsize=3
    )

plt.ylim(0, 1)
plt.tight_layout()
plt.savefig(os.path.join(exp_root, "plots", "exp3_limit.png"), dpi=300)

# RMSE
data_bl2 = data2.groupby(["Story", "Emotion"])["Val"].mean().reset_index()
data_bl2["Val"] = data_bl2.groupby("Story")["Val"].transform(lambda x: x / x.sum())

data_bl2 = data_bl2.merge(data_m2, on=["Story", "Emotion"], suffixes=("_x", "_y"))
data_bl2["d"] = (data_bl2["Val_x"] - data_bl2["Val_y"])**2

print("LIMIT RMSE:", np.sqrt(data_bl2["d"].mean()))

File: Exp3/04_statistical_analysis/Exp3_analyse.R
## Analyse the experiments of the RL emotion study
library(tidyverse)
library(lme4)
library(lmerTest)

args <- commandArgs(trailingOnly = FALSE)
file_arg <- grep("^--file=", args, value = TRUE)
script_dir <- if (length(file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", file_arg[1]), winslash = "/", mustWork = TRUE))
} else {
  getwd()
}
exp_root <- normalizePath(file.path(script_dir, ".."), winslash = "/", mustWork = TRUE)

source(file.path(script_dir, "theme-publication.R"))



######################
# free rating #
######################

data <- read.table(file.path(exp_root, "data", "human_free_limit.csv"), header = T, sep = ";") %>%
  pivot_longer(cols = starts_with("Emo."),
               names_to = "Emotion",
               names_prefix = "Emo.",
               values_to = "Val")

## test the hypothesis
model <- lmer(Val ~ Story*Emotion + (1|ID), data = data)
anova(model)

## Get 95% CI (approximate, not estimated)
data.h <- data %>% group_by(Story, Emotion) %>%
  summarise(sd = sd(Val),Val = mean(Val)) %>%
  mutate(SE = sd / sqrt(length(unique(data$ID))),
         CI_lower = Val - 1.96*SE,
         CI_upper = Val + 1.96*SE) %>%
  dplyr::select(Story, Emotion, Val, CI_lower, CI_upper)

data.m <- read.table(file.path(exp_root, "data", "svm_free_0.0013_var.csv"), header = T, sep = ",") %>%
  group_by(Story, Emotion) %>%
  summarise(Val = mean(Val)) %>%
  mutate(CI_lower = Val, CI_upper = Val)

## Visualise the effect

# ggplot(data %>% group_by(Story, Emotion) %>% summarise(Val = mean(Val)),
#        aes(Emotion, Val)) +
#   geom_bar(stat = "identity") +
#   facet_grid(Story ~ .) +
#   ylim(0,1)


## Visualise both data1 and data.m
# data.b <- data %>% group_by(Story, Emotion) %>% summarise(Val = mean(Val)) %>%
#   mutate(Val = Val/sum(Val)) %>%
#   mutate(Source = "Human") %>%
#   rbind(data.m %>% mutate(Source = "Model"))

data.b <- data.h %>%
  mutate(Source = "Human") %>%
  rbind(data.m %>% mutate(Source = "Model"))


ggplot(data.b, aes(Emotion, Val, fill = Source)) +
  geom_bar(stat = "identity",width = 0.8, position = position_dodge(0.85)) +
  geom_errorbar(aes(ymin = ifelse(Source == "Human", CI_lower, NA), 
                    ymax = ifelse(Source == "Human", CI_upper, NA)), 
                width = 0.2, position = position_dodge(0.85)) +  
  facet_grid(Story ~ .) +
  ylab("Intensity") + xlab('Emotion')+
  scale_y_continuous(limits = c(0,0.6),breaks = c(0,0.3,0.6))+
  theme_Publication() + scale_fill_Publication() + 
  theme(legend.position = c(0.92, 0.96),
        legend.margin = margin(1, 1, 1,1, "mm"),
        legend.title=element_blank(),
        legend.background = element_rect(fill='transparent'),
        panel.grid.major.x = element_blank() ,
        legend.key.size = unit(0.9, "lines"),
        legend.box.background = element_rect(colour = "black",fill="transparent"),
        legend.spacing.y = unit(0, "mm"))

ggsave(file = file.path(exp_root, "plots", "exp3_free.jpg"), width=8, height=8)

## Both data, long form
data.bl <- data %>% group_by(Story, Emotion) %>% summarise(Val = mean(Val)) %>%
  mutate(Val = Val/sum(Val)) %>%
  left_join(data.m, by = c("Story","Emotion")) %>%
  mutate(d = (Val.x - Val.y)^2)

## Investigate fit
ggplot(data.bl %>% group_by(C) %>% summarise(d = mean(d)) %>%
         mutate(d = sqrt(d)),
       aes(C, d)) +
  geom_point() +
  geom_smooth()


summary(lm(Val.x ~ Val.y, data = data.bl))

sqrt(data.bl%>%ungroup() %>%summarise(d=mean(d)))

#####################
# limit rating #
#####################



data2 <- read.table(file.path(exp_root, "data", "human_free_limit.csv"), header = T, sep = ";") %>%
  pivot_longer(cols = starts_with("mc"),
               names_to = "Emotion",
               names_prefix = "mc.",
               values_to = "Val")

## test the hypothesis
model <- lmer(Val ~ Story*Emotion + (1|ID), data = data2)
anova(model)

## Get 95% CI (approximate, not estimated)
data.h2 <- data2 %>% group_by(Story, Emotion) %>%
  summarise(sd = sd(Val),Val = mean(Val)) %>%
  mutate(SE = sd / sqrt(length(unique(data$ID))),
         CI_lower = pmax(Val - 1.96*SE, 0),
         CI_upper = pmin(Val + 1.96*SE, 1)) %>%
  dplyr::select(Story, Emotion, Val, CI_lower, CI_upper)

data.m2 <- read.table(file.path(exp_root, "data", "svm_limit_0.0034_var.csv"), 
                      header = T, sep = ",") %>%
  group_by(Story, Emotion) %>%
  summarise(Val = mean(Val))

# ggplot(data.m2 %>% group_by(Story, Emotion) %>% summarise(Val = mean(Val))%>%
#          mutate(Val = Val/sum(Val)),
#        aes(Emotion, Val)) +
#   geom_bar(stat = "identity") +
#   facet_grid(Story ~ .) +
#   ylim(0,1)

## Visualise both data2 and data.m
# data.b2 <- data2 %>% group_by(Story, Emotion) %>% summarise(Val = mean(Val)) %>%
#   mutate(Val = Val/sum(Val)) %>%
#   mutate(Source = "Human") %>%
#   rbind(data.m2 %>% mutate(Source = "Model"))

data.b2 <- data.h2 %>%
  mutate(Source = "Human") %>%
  rbind(data.m2 %>% mutate(Source = "Model"))

ggplot(data.b2, aes(Emotion, Val, fill = Source)) +
  geom_bar(stat = "identity",width = 0.8, position = position_dodge(0.85)) +
  geom_errorbar(aes(ymin = ifelse(Source == "Human", CI_lower, NA), 
                    ymax = ifelse(Source == "Human", CI_upper, NA)), 
                width = 0.2, position = position_dodge(0.85)) + 
  facet_grid(Story ~ .) +
  ylab("Intensity") + xlab('Emotion')+
  scale_y_continuous(limits = c(0,1),breaks = c(0,0.5,1))+
  theme_Publication() + scale_fill_Publication() + 
  theme(legend.position = c(0.92, 0.96),
        legend.margin = margin(1, 1, 1,1, "mm"),
        legend.title=element_blank(),
        legend.background = element_rect(fill='transparent'),
        panel.grid.major.x = element_blank() ,
        legend.key.size = unit(0.9, "lines"),
        legend.box.background = element_rect(colour = "black",fill="transparent"),
        legend.spacing.y = unit(0, "mm"))

ggsave(file = file.path(exp_root, "plots", "exp3_limit.jpg"), width=8, height=8)

## Both data, long form
data.bl2 <- data2 %>% group_by(Story, Emotion) %>% summarise(Val = mean(Val)) %>%
  mutate(Val = Val/sum(Val)) %>%
  left_join(data.m2, by = c("Story","Emotion")) %>%
  mutate(d = (Val.x - Val.y)^2)

## Investigate fit
ggplot(data.bl2 %>% group_by(C) %>% summarise(d = mean(d)) %>%
         mutate(d = sqrt(d)),
       aes(C, d)) +
  geom_point() +
  geom_smooth()


summary(lm(Val.x ~ Val.y, data = data.bl2))
sqrt(data.bl2%>%ungroup() %>%summarise(d=mean(d)))
